File: Moving_Average_from_DataStream/MovingAverage.py
import Queue

class MovingAverage(object):
    def __init__(self, size):
        global sum
        sum = 0.0
        self.size = size
        self.stream = Queue.Queue()

    def next(self, val):
        global sum
        if self.stream.qsize() == self.size:
            sum -= self.stream.get()
        sum += val
        self.stream.put(val)
        return sum / self.stream.qsize()

if __name__ == "__main__":
    m = MovingAverage(3)
    print "m.next(1) = %f\n" % m.next(1)
    print "m.next(10) = %f\n" % m.next(10)
    print "m.next(3) = %f\n" % m.next(3)
    print "m.next(5) = %f\n" % m.next(5)

File: Moving_Average_from_DataStream/MovingAverage.cpp
#include <iostream>
#include <queue>

class MovingAverage {
public:
    MovingAverage(int size) : _size(size) {}
    double next(int val) {
        if (stream.size() == _size) {
            sum -= stream.front();
            stream.pop();
        }
        sum += val;
        stream.push(val);
        return sum / stream.size();
    }
private:
    std::queue<int> stream;
    int _size;
    double sum = 0;
};

int main() {
    MovingAverage ma(3);
    std::cout << "ma.next(1) = " << ma.next(1) << std::endl;
    std::cout << "ma.next(10) = " << ma.next(10) << std::endl;
    std::cout << "ma.next(3) = " << ma.next(3) << std::endl;
    std::cout << "ma.next(5) = " << ma.next(5) << std::endl;
}
